### ISSUE TITLE
Diagnose misuse of the cleanup attribute

### ISSUE THREAD

**Issue by AaronBallman** [llvm/llvm-project#79443]
The cleanup attribute is written on a variable declaration, and the address of the declared variable is passed to the specified function in the attribute. This has some subtle issues we should diagnose on. Consider:
```
__attribute__((cleanup(free))) void *p = malloc(10);
```
This will pass `&p` to `free`, which is going to do the wrong thing. GCC diagnoses this kind of misuse (`warning: 'free' called on unallocated object 'p' [-Wfree-nonheap-object]`) and it would be good if we did the same: https://godbolt.org/z/nhqh9EazM

**Comment by 11happy** [comment]
Hello @AaronBallman I am interested to work on this , I have worked out a local fix for above issue but that's specific to `free`,
However I observed that this warning will be emitted with many sorts of standard functions like `closedir` , `fclose` etc. so how should I proceed , your guidance will be very helpful.
Thank you

**Comment by 11happy** [comment]
my local fix assumes that directly calling ``` __attribute__(cleanup(free)))``` will always give warning, I assume correct usage is like this way:
```
void cleanup_function(int *ptr) {
    free(ptr);
}

int main() {
    int *p __attribute__((cleanup(cleanup_function))) = malloc(sizeof(int));
```

**Comment by AaronBallman** [comment]
You're correct about what the correct usage should be. We don't want the logic to be specific to `free` though. My intuition is that the issue is somewhere in `Sema::CheckAssignmentConstraints()` which is called by `handleCleanupAttr()` in `SemaDeclAttr.cpp`; `void **` to `void *` is a compatible assignment in C but is not something we want in this particular case.
